fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,629 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+
+        self.in_queue.put(task)
+
+        ok, payload = self.out_queue.get()
+
+        # Hold the first exception.
+        if exc_info is not None:
+            pass
+        elif not ok:
+            exc_info = cast(ExcInfo, payload)
+        else:
+            task, batch = cast(Tuple[Task, Batch], payload)
+
+            with use_device(self.device):
+                task.finalize(batch)
+
+            pipeline_record.batches[chunk] = batch
+
+        if exc_info is not None:
+            raise exc_info[0].with_traceback(exc_info[1], exc_info[2])
+
+    def forward_results(self, chunk: int, pipeline_record: DistributedPipelineRecord) -> None:
+        """Forwards outputs of processing a chunk in this parition for processing by next partition."""
+        with use_stream(self.stream):
+            for user, input_idx, output_idx in pipeline_record.users:
+                v = pipeline_record.get_batch(chunk).value[output_idx]
+                pipeline_record.forwarded_phony[chunk][output_idx].append(user.remote().feed(chunk, input_idx, v))
+
+    def run_pipeline(self, pipeline_record_rref: rpc.RRef) -> Optional[Tensor]:
+        """Processes a min-batch on this partition.
+           If this is the last partition (pipeline_record has no user), concatenates results of processing
+           all chunks and returns the result as the output of the model on the whole mini-batch.
+        """
+        pipeline_record = pipeline_record_rref.local_value()
+        self.run(pipeline_record)
+
+        if not pipeline_record.users:
+            result = microbatch.gather(pipeline_record.batches)
+            assert len(result) == 1
+            result = result[0]
+            s0 = current_stream(result.device)
+            if is_cuda(s0):
+                # TODO. Investigate why this is needed and remove it if possible.
+                as_cuda(s0).synchronize()
+            return result
+
+        return None
+
+
+class MultiInputSequential(nn.Module):
+    """A variation of nn.Sequential, that allows the first module in the sequence accepts
+        multiple inputs. To be used internally by _split_module
+    """
+
+    def __init__(self, *modules: nn.Module) -> None:
+        super().__init__()
+        self.modules_list = modules
+
+    def forward(self, *inputs: Tuple[Tensor]) -> Tensor:  # type: ignore
+        input = self.modules_list[0](*inputs)
+        for module in self.modules_list[1:]:
+            input = module(input)
+        return input
+
 
-        model = nn.Sequential(a, b, c, d)
-        model = Pipe(model, balance=[1, 1, 1, 1], chunks=8)
-        output = model(input)
+def RemoteSequential(rref_list: List[rpc.RRef]) -> MultiInputSequential:
+    return MultiInputSequential(*(r.local_value() for r in rref_list))
 
-    .. _Pipe: https://arxiv.org/abs/1811.06965
 
-    Pipe combines pipeline parallelism with checkpointing to reduce peak
+def _split_module(graph: PipelineModulesGraph) -> List[Tuple[List[int], rpc.RRef]]:
```

Comment:
There are a lot of layering issues with this function. It is accessing many internal properties of PipelineModulesGraph. To understand requires going back to the PipelineModulesGraph. Should this be a method of PipelineModulesGraph. Or maybe some additional methods should be added to PipelineModulesGraph to that hide some of the details exposed here.